src/clj/puppetlabs/services/jruby_pool_manager/impl/jruby_agents.clj

```diff
@@ -86,47 +87,67 @@
    pool :- jruby-schemas/pool-queue-type]
   (= expected-pool-size (count (.getRegisteredElements pool))))
 
-(schema/defn ^:always-validate
-  swap-and-drain-pool!
-  "Replace the current pool with a new pool and drain the old pool,
-  optionally refilling the new pool with fresh jrubies."
+(schema/defn collect-all-jrubies :- [JRubyInstance]
+  "Locks the pool and borrows all the instances"
+  [pool-state :- jruby-schemas/PoolState]
+  (let [pool-size (:size pool-state)
+        pool (:pool pool-state)
+        borrow-fn (fn [_] (jruby-internal/borrow-from-pool!*
+                           jruby-internal/borrow-without-timeout-fn
+                           pool))]
+    (.lock pool)
+    (try
+      (mapv borrow-fn (range pool-size))
+      (catch Exception e
+        (.clear pool)
+        (.insertPill pool (PoisonPill. e))
+        (throw (IllegalStateException.
+                "There was a problem borrowing a JRubyInstance from the pool."
+                e)))
+      (finally
+        (.unlock pool)))))
+
+(schema/defn cleanup-and-refill-pool
+  "Cleans up the given instances and optionally refills the pool with
+  new instances. Should only be called from the modify-instance-agent"
   [pool-context :- jruby-schemas/PoolContext
-   old-pool-state :- jruby-schemas/PoolState
-   new-pool-state :- jruby-schemas/PoolState
+   pool-state :- jruby-schemas/PoolState
+   old-instances :- [JRubyInstance]
    refill? :- schema/Bool]
-  (let [{:keys [config]} pool-context
-        pool-state-atom (jruby-internal/get-pool-state-container pool-context)
-        new-pool (:pool new-pool-state)
-        old-pool (:pool old-pool-state)
-        old-pool-size (:size old-pool-state)
+  (let [pool (:pool pool-state)
+        pool-size (:size pool-state)
+        new-instance-ids (map inc (range pool-size))
+        config (:config pool-context)
         cleanup-fn (get-in config [:lifecycle :cleanup])]
-    (log/info "Replacing old JRuby pool with new instance.")
-    (reset! pool-state-atom new-pool-state)
-    (log/info "Swapped JRuby pools, beginning cleanup of old pool.")
-    (doseq [i (range old-pool-size)]
-      (try
-        (let [id (inc i)
-              instance (jruby-internal/borrow-from-pool!*
-                        jruby-internal/borrow-without-timeout-fn
-                        old-pool)]
-          (try
-            (jruby-internal/cleanup-pool-instance! instance cleanup-fn)
-            (when refill?
-              (jruby-internal/create-pool-instance! new-pool id config
-                                                    (partial send-flush-instance! pool-context))
-              (log/infof "Finished creating JRubyInstance %d of %d"
-                         id old-pool-size))
-            (finally
-              (.releaseItem old-pool instance false))))
```

Comment:
looks like it's still being used in the `jruby-internal/return-to-pool` function so that an instance won't be returned to the pool if an exception occurs while flushing it